add speaker recognition recipe for aishell

naxingyu opened the pull request.

danpovey commented:
@david-ryan-snyder, can you please review this?

david-ryan-snyder commented:
@naxingyu, @danpovey, sure, will be happy to help -- I will get back to this in a few days.

david-ryan-snyder commented on egs/aishell/v1/run.sh:
```diff
@@ -0,0 +1,92 @@
+# This is a shell script, but it's recommended that you run the commands one by
+# one by copying and pasting into the shell.
+# Caution: some of the graph creation steps use quite a bit of memory, so you
+# should run this on a machine that has sufficient memory.
```
The comment about graph creation is out of date. You might want to have another comment here that reads something like:

`# See README.txt for more info on data required.`
`# Results (mostly EERs) are inline in comments below.`

david-ryan-snyder replied:
And, please add a README.txt if there isn't one already.

david-ryan-snyder commented on egs/aishell/v1/run.sh:
```diff
@@ -0,0 +1,92 @@
+#train plda
+ivector-compute-plda ark:data/train/spk2utt \
+  'ark:ivector-normalize-length scp:exp/ivector_train_1024/ivector.scp  ark:- |' \
+  exp/ivector_train_1024/plda 2>exp/ivector_train_1024/log/plda.log
```
For consistency with other newer scripts, could you invoke the binary using $train_cmd, like this: https://github.com/kaldi-asr/kaldi/blob/master/egs/sre16/v1/run.sh#L220

david-ryan-snyder commented on egs/aishell/v1/run.sh:
```diff
@@ -0,0 +1,92 @@
+mkdir -p data/test/enroll data/test/match
+cp data/test/{spk2utt,feats.scp,vad.scp}  data/test/enroll
+cp data/test/{spk2utt,feats.scp,vad.scp}  data/test/match
+python local/split_data_enroll_match.py data/test/utt2spk  data/test/enroll/utt2spk  data/test/match/utt2spk
```
Please remove extra spaces between the previous arguments. Also, you can call the python script without explicitly writing "python" as you've added the shebang line to your python scripts.

david-ryan-snyder commented on egs/aishell/v1/run.sh:
```diff
@@ -0,0 +1,92 @@
+python local/split_data_enroll_match.py data/test/utt2spk  data/test/enroll/utt2spk  data/test/match/utt2spk
+python local/produce_trials.py data/test/enroll/utt2spk data/test/match/utt2spk data/test/aishell_speaker_ver.lst
+utils/fix_data_dir.sh data/test/enroll
+utils/fix_data_dir.sh data/test/match
```
Also remove the explicit use of "python" here

david-ryan-snyder commented on egs/aishell/v1/run.sh:
```diff
@@ -0,0 +1,92 @@
+	exp/ivector_train_1024/plda \
+	ark:exp/ivector_enroll_1024/spk_ivector.ark \
+	"ark:ivector-normalize-length scp:exp/ivector_match_1024/ivector.scp ark:- |" \
+	"cat '$trials' | awk '{print \$2, \$1}' |" exp/trials_out || exit 1
```
The indent level for lines 71-83 isn't consistent with the rest of the file, which uses to spaces. Also, please run the binary using $train_cmd as is done in https://github.com/kaldi-asr/kaldi/blob/master/egs/sre16/v1/run.sh#L220.

david-ryan-snyder commented on egs/aishell/v1/local/produce_trials.py:
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env python
```
I believe that @danpovey now prefers python3 for new python scripts. Could you change this to read "python3" and then verify that the script still works?

david-ryan-snyder commented on egs/aishell/v1/local/produce_trials.py:
```diff
@@ -0,0 +1,23 @@
+            trial = utt2 + ' ' + spk3 + ' ' + 'target'
+        else:
+            trial = utt2 + ' ' + spk3 + ' ' + 'nontarget'
+        trailfile.write(trial+'\n')
```
I think Kaldi usually uses 2 spaces for indent, even in python scripts.

david-ryan-snyder commented on egs/aishell/v1/local/split_data_enroll_match.py:
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env python
```
Should be python3 now.

david-ryan-snyder commented on egs/aishell/v1/local/split_data_enroll_match.py:
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env python
+
+# Copyright 2017 Bengu Wu
+# Apache 2.0.
```
I think it would be nice to leave a comment describing what this script is doing.

david-ryan-snyder commented on egs/aishell/v1/conf/mfcc.conf:
```diff
@@ -0,0 +1,2 @@
+--use-energy=false   # only non-default option.
```
This looks funny to me... Does it still make sense to use the energy VAD if --use-energy=false? @danpovey, could you comment on this one?

david-ryan-snyder commented on egs/aishell/v1/local/produce_trials.py:
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env python
+
+# Copyright 2017 Bengu Wu
+# Apache 2.0.
```
Since this script defines the evaluation you're going to perform, it would be great to get some comment here, describing how the trials are generated.

naxingyu commented:
Thank you. This recipe was created referring to an out-of-date sre example.
We will update accordingly.

2018-01-07 1:29 GMT+08:00 david-ryan-snyder <notifications@github.com>:

> *@david-ryan-snyder* commented on this pull request.
>
> @naxingyu <https://github.com/naxingyu>, I left a few suggestions.
> ------------------------------
>
> In egs/aishell/v1/run.sh
> <https://github.com/kaldi-asr/kaldi/pull/2111#discussion_r160030108>:
>
> > @@ -0,0 +1,92 @@
> +#!/bin/bash
> +
> +# Copyright 2017 Beijing Shell Shell Tech. Co. Ltd. (Authors: Hui Bu)
> +#           2017 Jiayu Du
> +#           2017 Chao Li
> +#           2017 Xingyu Na
> +#           2017 Bengu Wu
> +#           2017 Hao Zheng
> +# Apache 2.0
> +
> +# This is a shell script, but it's recommended that you run the commands one by
> +# one by copying and pasting into the shell.
> +# Caution: some of the graph creation steps use quite a bit of memory, so you
> +# should run this on a machine that has sufficient memory.
>
> The comment about graph creation is out of date. You might want to have
> another comment here that reads something like:
>
> # See README.txt for more info on data required.
> # Results (mostly EERs) are inline in comments below.
> ------------------------------
>
> In egs/aishell/v1/run.sh
> <https://github.com/kaldi-asr/kaldi/pull/2111#discussion_r160030174>:
>
> > +sid/train_full_ubm.sh --nj 10 --cmd "$train_cmd" data/train \
> +  exp/diag_ubm_1024 exp/full_ubm_1024
> +
> +#train ivector
> +sid/train_ivector_extractor.sh --cmd "$train_cmd" \
> +  --num-iters 5 exp/full_ubm_1024/final.ubm data/train \
> +  exp/extractor_1024
> +
> +#extract ivector
> +sid/extract_ivectors.sh --cmd "$train_cmd" --nj 10 \
> +  exp/extractor_1024 data/train exp/ivector_train_1024
> +
> +#train plda
> +ivector-compute-plda ark:data/train/spk2utt \
> +  'ark:ivector-normalize-length scp:exp/ivector_train_1024/ivector.scp  ark:- |' \
> +  exp/ivector_train_1024/plda 2>exp/ivector_train_1024/log/plda.log
>
> For consistency with other newer scripts, could you invoke the binary
> using $train_cmd, like this: https://github.com/kaldi-asr/
> kaldi/blob/master/egs/sre16/v1/run.sh#L220
> ------------------------------
>
> In egs/aishell/v1/run.sh
> <https://github.com/kaldi-asr/kaldi/pull/2111#discussion_r160030253>:
>
> > +  exp/extractor_1024
> +
> +#extract ivector
> +sid/extract_ivectors.sh --cmd "$train_cmd" --nj 10 \
> +  exp/extractor_1024 data/train exp/ivector_train_1024
> +
> +#train plda
> +ivector-compute-plda ark:data/train/spk2utt \
> +  'ark:ivector-normalize-length scp:exp/ivector_train_1024/ivector.scp  ark:- |' \
> +  exp/ivector_train_1024/plda 2>exp/ivector_train_1024/log/plda.log
> +
> +#split the test to enroll and match
> +mkdir -p data/test/enroll data/test/match
> +cp data/test/{spk2utt,feats.scp,vad.scp}  data/test/enroll
> +cp data/test/{spk2utt,feats.scp,vad.scp}  data/test/match
> +python local/split_data_enroll_match.py data/test/utt2spk  data/test/enroll/utt2spk  data/test/match/utt2spk
>
> Please remove extra spaces between the previous arguments. Also, you can
> call the python script without explicitly writing "python" as you've added
> the shebang line to your python scripts.
> ------------------------------
>
> In egs/aishell/v1/run.sh
> <https://github.com/kaldi-asr/kaldi/pull/2111#discussion_r160030270>:
>
> > +sid/extract_ivectors.sh --cmd "$train_cmd" --nj 10 \
> +  exp/extractor_1024 data/train exp/ivector_train_1024
> +
> +#train plda
> +ivector-compute-plda ark:data/train/spk2utt \
> +  'ark:ivector-normalize-length scp:exp/ivector_train_1024/ivector.scp  ark:- |' \
> +  exp/ivector_train_1024/plda 2>exp/ivector_train_1024/log/plda.log
> +
> +#split the test to enroll and match
> +mkdir -p data/test/enroll data/test/match
> +cp data/test/{spk2utt,feats.scp,vad.scp}  data/test/enroll
> +cp data/test/{spk2utt,feats.scp,vad.scp}  data/test/match
> +python local/split_data_enroll_match.py data/test/utt2spk  data/test/enroll/utt2spk  data/test/match/utt2spk
> +python local/produce_trials.py data/test/enroll/utt2spk data/test/match/utt2spk data/test/aishell_speaker_ver.lst
> +utils/fix_data_dir.sh data/test/enroll
> +utils/fix_data_dir.sh data/test/match
>
> Also remove the explicit use of "python" here
> ------------------------------
>
> In egs/aishell/v1/run.sh
> <https://github.com/kaldi-asr/kaldi/pull/2111#discussion_r160030353>:
>
> > @@ -0,0 +1,92 @@
> +#!/bin/bash
> +
> +# Copyright 2017 Beijing Shell Shell Tech. Co. Ltd. (Authors: Hui Bu)
> +#           2017 Jiayu Du
> +#           2017 Chao Li
> +#           2017 Xingyu Na
> +#           2017 Bengu Wu
> +#           2017 Hao Zheng
> +# Apache 2.0
> +
> +# This is a shell script, but it's recommended that you run the commands one by
> +# one by copying and pasting into the shell.
> +# Caution: some of the graph creation steps use quite a bit of memory, so you
> +# should run this on a machine that has sufficient memory.
>
> And, please add a README.txt if there isn't one already.
> ------------------------------
>
> In egs/aishell/v1/run.sh
> <https://github.com/kaldi-asr/kaldi/pull/2111#discussion_r160030361>:
>
> > +utils/fix_data_dir.sh data/test/match
> +
> +#extract enroll ivector
> +sid/extract_ivectors.sh --cmd "$train_cmd" --nj 10 \
> +	exp/extractor_1024 data/test/enroll  exp/ivector_enroll_1024
> +#extract match ivector
> +sid/extract_ivectors.sh --cmd "$train_cmd" --nj 10 \
> +	exp/extractor_1024 data/test/match  exp/ivector_match_1024
> +
> +#compute plda score
> +trials=data/test/aishell_speaker_ver.lst
> +ivector-plda-scoring --num-utts=ark:exp/ivector_enroll_1024/num_utts.ark \
> +	exp/ivector_train_1024/plda \
> +	ark:exp/ivector_enroll_1024/spk_ivector.ark \
> +	"ark:ivector-normalize-length scp:exp/ivector_match_1024/ivector.scp ark:- |" \
> +	"cat '$trials' | awk '{print \$2, \$1}' |" exp/trials_out || exit 1
>
> The indent level for lines 71-83 isn't consistent with the rest of the
> file, which uses to spaces. Also, please run the binary using $train_cmd as
> is done in https://github.com/kaldi-asr/kaldi/blob/master/egs/sre16/
> v1/run.sh#L220.
> ------------------------------
>
> In egs/aishell/v1/local/produce_trials.py
> <https://github.com/kaldi-asr/kaldi/pull/2111#discussion_r160030393>:
>
> > @@ -0,0 +1,23 @@
> +#!/usr/bin/env python
>
> I believe that @danpovey <https://github.com/danpovey> now prefers
> python3 for new python scripts. Could you change this to read "python3" and
> then verify that the script still works?
> ------------------------------
>
> In egs/aishell/v1/local/produce_trials.py
> <https://github.com/kaldi-asr/kaldi/pull/2111#discussion_r160030435>:
>
> > +dictutt = {}
> +for eachline in open(sys.argv[1]):
> +    line = eachline.rstrip('\r\t\n ')
> +    spk = line.split(' ')[1]
> +    dictutt[spk] = spk
> +
> +trailfile = open(sys.argv[3], 'w')
> +for each2 in open(sys.argv[2]):
> +    line2 = each2.rstrip('\r\t\n ')
> +    utt2, spk2 = line2.split(' ')
> +    for spk3 in dictutt:
> +        if spk3 == spk2:
> +            trial = utt2 + ' ' + spk3 + ' ' + 'target'
> +        else:
> +            trial = utt2 + ' ' + spk3 + ' ' + 'nontarget'
> +        trailfile.write(trial+'\n')
>
> I think Kaldi usually uses 2 spaces for indent, even in python scripts.
> ------------------------------
>
> In egs/aishell/v1/local/split_data_enroll_match.py
> <https://github.com/kaldi-asr/kaldi/pull/2111#discussion_r160030455>:
>
> > @@ -0,0 +1,34 @@
> +#!/usr/bin/env python
>
> Should be python3 now.
> ------------------------------
>
> In egs/aishell/v1/local/split_data_enroll_match.py
> <https://github.com/kaldi-asr/kaldi/pull/2111#discussion_r160030471>:
>
> > @@ -0,0 +1,34 @@
> +#!/usr/bin/env python
> +
> +# Copyright 2017 Bengu Wu
> +# Apache 2.0.
>
> I think it would be nice to leave a comment describing what this script is
> doing.
> ------------------------------
>
> In egs/aishell/v1/conf/mfcc.conf
> <https://github.com/kaldi-asr/kaldi/pull/2111#discussion_r160030525>:
>
> > @@ -0,0 +1,2 @@
> +--use-energy=false   # only non-default option.
>
> This looks funny to me... Does it still make sense to use the energy VAD
> if --use-energy=false? @danpovey <https://github.com/danpovey>, could you
> comment on this one?
> ------------------------------
>
> In egs/aishell/v1/local/produce_trials.py
> <https://github.com/kaldi-asr/kaldi/pull/2111#discussion_r160030585>:
>
> > @@ -0,0 +1,23 @@
> +#!/usr/bin/env python
> +
> +# Copyright 2017 Bengu Wu
> +# Apache 2.0.
>
> Since this script defines the evaluation you're going to perform, it would
> be great to get some comment here, describing how the trials are generated.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/2111#pullrequestreview-87079824>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADKpxFKITisxhim4SPpBciUM5bALUm6Qks5tH62XgaJpZM4RO1rE>
> .
>



-- 
Xingyu Na

naxingyu commented:
@david-ryan-snyder All comments are resolved. Please review. Thank you!

naxingyu commented:
@david-ryan-snyder Updated. The term 'match' is replaced with 'eval'. We don't call it 'test' because it is confusing considering there are train/dev/test sets in the original AISHELL-1 data.

david-ryan-snyder commented:
Thanks @naxingyu.

Sorry to nitpick, but it looked like you were in the process of removing the extra "|| exit 1" in the run.sh, but forgot to remove some of them (when you're making mfccs and fixing the data dir).

naxingyu commented on egs/aishell/v1/run.sh:
```diff
@@ -0,0 +1,93 @@
+  steps/make_mfcc.sh --cmd "$train_cmd" --nj 10 data/$x exp/make_mfcc/$x $mfccdir || exit 1;
+  steps/compute_cmvn_stats.sh data/$x exp/make_mfcc/$x $mfccdir || exit 1;
+  sid/compute_vad_decision.sh --nj 10 --cmd "$train_cmd" data/$x exp/make_mfcc/$x $mfccdir
+  utils/fix_data_dir.sh data/$x || exit 1;
```
Oh, I remember there were '|| exit 1's in run.sh scripts. When are they removed? How does run.sh prevent it from continuing running the next command if a script fails?

danpovey commented:
You can add `set -e` somewhere near the top.


On Sat, Jan 20, 2018 at 5:29 AM, Xingyu Na <notifications@github.com> wrote:

> *@naxingyu* commented on this pull request.
> ------------------------------
>
> In egs/aishell/v1/run.sh
> <https://github.com/kaldi-asr/kaldi/pull/2111#discussion_r162779479>:
>
> > +
> +local/download_and_untar.sh $data $data_url data_aishell || exit 1;
> +local/download_and_untar.sh $data $data_url resource_aishell || exit 1;
> +
> +# Data Preparation
> +local/aishell_data_prep.sh $data/data_aishell/wav $data/data_aishell/transcript || exit 1;
> +
> +# Now make MFCC  features.
> +# mfccdir should be some place with a largish disk where you
> +# want to store MFCC features.
> +mfccdir=mfcc
> +for x in train test; do
> +  steps/make_mfcc.sh --cmd "$train_cmd" --nj 10 data/$x exp/make_mfcc/$x $mfccdir || exit 1;
> +  steps/compute_cmvn_stats.sh data/$x exp/make_mfcc/$x $mfccdir || exit 1;
> +  sid/compute_vad_decision.sh --nj 10 --cmd "$train_cmd" data/$x exp/make_mfcc/$x $mfccdir
> +  utils/fix_data_dir.sh data/$x || exit 1;
>
> Oh, I remember there were '|| exit 1's in run.sh scripts. When are they
> removed? How does run.sh prevent it from continuing running the next
> command if a script fails?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/2111#discussion_r162779479>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu2IiZK-TuJyc6_RtXMiEfTBtNC5qks5tMcACgaJpZM4RO1rE>
> .
>

naxingyu commented:
Done. @david-ryan-snyder

david-ryan-snyder commented:
Thanks @naxingyu.

@danpovey, I think this should be OK to merge now.
